File: android/localasset-core/src/main/kotlin/io/github/xesam/android/localasset/core/engine/DefaultLocalAssetSchemeAdapter.kt
package io.github.xesam.android.localasset.core.engine

import io.github.xesam.android.localasset.core.api.SchemeAdapter
import io.github.xesam.android.localasset.core.error.ResourceErrorCategory
import io.github.xesam.android.localasset.core.error.ResourceException
import io.github.xesam.android.localasset.core.model.AssetRequest
import java.net.URI
import java.net.URLDecoder

class DefaultLocalAssetSchemeAdapter : SchemeAdapter {
    override fun canHandle(url: String): Boolean = runCatching {
        URI(url).scheme == "local-asset"
    }.getOrDefault(false)

    override fun parse(url: String): AssetRequest {
        val uri = runCatching { URI(url) }.getOrElse {
            throw ResourceException(ResourceErrorCategory.PARSE_ERROR, "invalid url")
        }
        if (uri.scheme != "local-asset") {
            throw ResourceException(ResourceErrorCategory.PARSE_ERROR, "unsupported scheme")
        }
        val namespace = uri.host?.takeIf { it.isNotBlank() }
            ?: throw ResourceException(ResourceErrorCategory.PARSE_ERROR, "missing namespace")
        val rawPath = uri.path?.takeIf { it.isNotBlank() }
            ?: throw ResourceException(ResourceErrorCategory.PARSE_ERROR, "missing path")
        val path = if (rawPath.startsWith("/")) rawPath else "/$rawPath"
        val identifier = path.substringAfterLast('/').takeIf { it.isNotBlank() }
        return AssetRequest(
            scheme = uri.scheme,
            namespace = namespace,
            identifier = identifier,
            path = path,
            query = parseQuery(uri.rawQuery),
            fragment = uri.fragment,
        )
    }

    private fun parseQuery(rawQuery: String?): Map<String, String> {
        if (rawQuery.isNullOrBlank()) return emptyMap()
        return rawQuery.split("&")
            .filter { it.isNotBlank() }
            .associate { part ->
                val pieces = part.split("=", limit = 2)
                decode(pieces[0]) to decode(pieces.getOrElse(1) { "" })
            }
    }

    private fun decode(value: String): String = URLDecoder.decode(value, "UTF-8")
}

File: ios/localasset-core/Sources/LocalAssetCore/engine/DefaultLocalAssetSchemeAdapter.swift
import Foundation

public class DefaultLocalAssetSchemeAdapter: SchemeAdapter {
    public init() {}

    public func canHandle(url: String) -> Bool {
        URLComponents(string: url)?.scheme == "local-asset"
    }

    public func parse(url: String) throws -> AssetRequest {
        guard let comps = URLComponents(string: url) else {
            throw ResourceException(.parseError, "invalid url")
        }
        guard comps.scheme == "local-asset" else {
            throw ResourceException(.parseError, "unsupported scheme")
        }
        guard let namespace = comps.host, !namespace.isEmpty else {
            throw ResourceException(.parseError, "missing namespace")
        }
        let rawPath = comps.path
        guard !rawPath.isEmpty else {
            throw ResourceException(.parseError, "missing path")
        }
        let path = rawPath.hasPrefix("/") ? rawPath : "/\(rawPath)"
        let identifier = path.split(separator: "/").last.map(String.init).flatMap { $0.isEmpty ? nil : $0 }
        let query = Dictionary(
            (comps.queryItems ?? []).compactMap { item -> (String, String)? in
                guard let value = item.value else { return nil }
                return (item.name, value)
            },
            uniquingKeysWith: { first, _ in first }
        )
        return AssetRequest(
            scheme: comps.scheme!,
            namespace: namespace,
            identifier: identifier,
            path: path,
            query: query,
            fragment: comps.fragment
        )
    }
}
